pandas/io/stata.py

```diff
@@ -2568,23 +2623,21 @@ def _prepare_data(self):
                     dtype = dtype.newbyteorder(self._byteorder)
                 dtypes[col] = dtype
 
-        self.data = data.to_records(index=False, column_dtypes=dtypes)
-
-    def _write_data(self):
-        data = self.data
-        self._file.write(data.tobytes())
+        return data.to_records(index=False, column_dtypes=dtypes)
 
-    def _null_terminate(self, s, as_string=False):
-        null_byte = "\x00"
-        s += null_byte
-
-        if not as_string:
-            s = s.encode(self._encoding)
+    def _write_data(self, records: np.recarray) -> None:
+        self._write_bytes(records.tobytes())
 
+    @staticmethod
+    def _null_terminate_str(s: str) -> str:
+        s += "\x00"
         return s
 
+    def _null_terminate_bytes(self, s: str) -> bytes:
+        return self._null_terminate_str(s).encode(self._encoding)
+
 
-def _dtype_to_stata_type_117(dtype, column, force_strl):
+def _dtype_to_stata_type_117(dtype: np.dtype, column: Series, force_strl: bool) -> int:
```

Comment:
Can you use `Dtype` from pandas._typing here?

Reply:
I'm not sure this is correct here since Extension dtypes are not supported.